tool/cargo/Syncer.kt

```diff
@@ -0,0 +1,407 @@
+
+
+import picocli.CommandLine
+import java.io.File
+import java.io.FileInputStream
+import java.nio.charset.StandardCharsets
+import java.nio.file.Files
+import java.nio.file.Path
+import java.util.*
+import java.util.concurrent.Callable
+import kotlin.io.path.Path
+import kotlin.system.exitProcess
+
+fun main(args: Array<String>): Unit = exitProcess(CommandLine(Syncer()).execute(*args))
+
+@CommandLine.Command(name = "sync", mixinStandardHelpOptions = true)
+class Syncer : Callable<Unit> {
+
+    @CommandLine.Option(names = ["--verbose", "-v"], required = false)
+    private var verbose: Boolean = false
+
+    private lateinit var logger: Logger
+    private lateinit var shell: Shell
+    private val workspaceDir = Path(System.getenv("BUILD_WORKSPACE_DIRECTORY"))
+    private lateinit var bazelOutputBase: Path
+
+    override fun call() {
+        logger = Logger(logLevel = if (verbose) DEBUG else ERROR)
+        shell = Shell(logger, verbose)
+        bazelOutputBase = Path(shell.execute(listOf(BAZEL, INFO, OUTPUT_BASE), workspaceDir).outputString().trim())
+
+        val rustTargets = rustTargets(shell, workspaceDir)
+        validateTargets(rustTargets)
+        loadRustToolchainAndExternalDeps(rustTargets)
+        vaticleRustWorkspaces().forEach { WorkspaceSyncer(it, logger, shell, bazelOutputBase).sync() }
+    }
+
+    private fun validateTargets(targets: List<String>) {
+        shell.execute(command = listOf(BAZEL, BUILD) + targets + "--build=false", baseDir = workspaceDir)
+    }
+
+    private fun loadRustToolchainAndExternalDeps(rustTargets: List<String>) {
+        shell.execute(command = listOf(BAZEL, BUILD) + rustTargets, baseDir = workspaceDir, throwOnError = false)
+    }
+
+    private fun vaticleRustWorkspaces(): List<Path> {
+        // e.g: [/Users/root/workspace/typedb-client-rust, /private/var/_bazel_root_/123abc/external/vaticle_typedb_protocol]
+        return listOf(workspaceDir) + shell.execute(listOf(BAZEL, QUERY, RUST_TARGETS_DEPS_QUERY, "--output=package"), workspaceDir)
+                .outputString().split("\n")
+                .filter { it.startsWith(VATICLE_REPOSITORY_PREFIX) }
+                .map { it.split("@")[1].split("//")[0] }
+                .map { bazelOutputBase.resolve("external").resolve(it) }
+    }
+
+    companion object {
+        private fun rustTargets(shell: Shell, workspace: Path): List<String> {
+            return shell.execute(listOf(BAZEL, QUERY, RUST_TARGETS_QUERY), workspace)
+                    .outputString().split("\n").filter { it.isNotBlank() }
+        }
+    }
+
+    private class WorkspaceSyncer(private val workspace: Path, private var logger: Logger, private var shell: Shell, var bazelOutputBase: Path) {
+
+        private val bazelBin = workspace.resolve(BAZEL_BIN).toRealPath().toFile()
+
+        fun sync() {
+            logger.debug { "Syncing $workspace" }
+            cleanupOldSyncProperties()
+            runSyncPropertiesAspect()
+            generateManifests()
```

Comment:
Now, instead of creating a new class here which handles the search of sync properties files, and then convertes each into manifests, the search of properties files is handled directly in this class. Each one is then converted to a Manifest with the `ManifestGenerator`